tutorials/grails3-send-push-notification-using-firebase-cloud-messaging

```diff
@@ -0,0 +1,206 @@
+2. Type this command.
+       
+   grails create-app grails-fcm
+       
+3. Go to the newly created Grails 3 application folder.
+    
+   cd grails-fcm
+    
+4. Enter Grails 3 interactive console by type this command.
+    
+   grails
+    
+5. In the Grails interactive console, you will see the following:
+    
+   | Resolving Dependencies. Please wait...
+   
+   CONFIGURE SUCCESSFUL
+
+   | Enter a command name to run. Use TAB for completion:
+   grails>
+    
+6. Test your Grails web application by typing this command in the Grails interactive console.
+    
+   run-app
+    
+## Add Grails Datastore Rest Client Dependency
+
+To access Firebase using REST API, add the Grails Datastore Client Rest dependency to the Grails 3 application.
+
+1. Open and edit build.gradle.
+
+2. Add the Grails Datastore Client Rest dependency inside the dependencies section. There are two dependencies, one inside buildscript and the other outside buildscripts.
+    
+dependencies {
+    ...
+    compile "org.hibernate:hibernate-core:5.1.2.Final"
+    compile "org.hibernate:hibernate-ehcache:5.1.2.Final"
+    compile 'org.grails:grails-datastore-rest-client:5.0.0.RC2'
+    console "org.grails:grails-console"
+    profile "org.grails.profiles:web"
+    runtime "com.bertramlabs.plugins:asset-pipeline-grails:2.11.6"
+    ...
+}
+    
+3. Compile the Grails 3 application by typing this command in the Grails interactive console.
+    
+compile
+    
+## Create Grails Controller and View for Sending Push Notification
+
+Run the following steps:
+
+1. Create controller by typing this command.
+    
+   create-controller Fcm
+    
+2. Add import to the file "grails-app/controllers/grails/fcm/FcmController.groovy".
+    
+   import grails.plugins.rest.client.RestBuilder
+    
+3. Create a method to FcmController, so it will looks like this.
+    
+    package grails.fcm
+
+    import grails.plugins.rest.client.RestBuilder
+
+    class FcmController {
+
+    def index() { }
+
+    def sendPushNotification() {
+     def regid = params.regid
+     def title = params.title
+     def body = params.body
+
+     def rest = new RestBuilder(connectTimeout:1000, readTimeout:20000)
+     def resp = rest.post("https://fcm.googleapis.com/fcm/send") {
+       header 'Content-Type', 'application/json'
+       header 'Authorization', 'key=AIza*****'
+       json {
+         notification = {
+           title = title
+           body = body
+           sound = "default"
+           click_action = "FCM_PLUGIN_ACTIVITY"
+           icon = "fcm_push_icon"
+         }
+         to = regid
+       }
+     }
+
+     flash.message = "Notification sent"
+     redirect action: "index"
+    }
+   }
+    
+   Enter the Legacy Server key you saved from your project dashboard in the line that begins with 'key='.
+    
+   header 'Authorization', 'key=AIza*****'
+    
+
+4. Create a view file "grails-app/views/fcm/index.gsp" and make this file like this.
+    
+    <!doctype html>
```

Comment:
[Here's an example](https://github.com/GoogleCloudPlatform/community/tree/master/tutorials/deploy-grails-to-google-cloud) of a tutorial where a folder + index.md file was used. That tutorial's `index.md` file originally had some HTML in it, but had to pull the html out into a [editFeaturedImage.gsp](https://github.com/GoogleCloudPlatform/community/blob/master/tutorials/deploy-grails-to-google-cloud/editFeaturedImage.gsp) file. That `index.md` now refers to that `.gsp` file with the following markdown:

```md
Create a file named `grails-app/views/book/editFeaturedImage.gsp` from the
    content found in [editFeaturedImage.gsp](https://github.com/GoogleCloudPlatform/community/blob/master/tutorials/deploy-grails-to-google-cloud/editFeaturedImage.gsp).
```